v2/spiffetls/listen.go

```diff
@@ -24,20 +26,74 @@ func Listen(ctx context.Context, network, laddr string, authorizer tlsconfig.Aut
 //
 // This function is the same as NewListenerWithMode using the MTLSServer mode.
 func NewListener(ctx context.Context, inner net.Listener, authorizer tlsconfig.Authorizer, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+	return NewListenerWithMode(ctx, inner, MTLSServer(authorizer), options...)
 }
 
 // ListenWithMode creates a TLS listener accepting connections on the given
 // network address using net.Listen. The given mode controls the authentication
 // and authorization exercised during the TLS handshake.
 func ListenWithMode(ctx context.Context, network, laddr string, mode ListenMode, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+	inner, err := net.Listen(network, laddr)
+	if err != nil {
+		return nil, err
+	}
+
+	l, err := NewListenerWithMode(ctx, inner, mode, options...)
+	if err != nil {
+		inner.Close()
+		return nil, err
+	}
+
+	return l, nil
 }
 
 // NewListenerWithMode creates a TLS listener which accepts connections from an
 // inner Listener and wraps each connection with tls.Server. The given mode
 // controls the authentication and authorization exercised during the TLS
 // handshake.
-func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (net.Listener, error) {
-	panic("not implemented")
+func NewListenerWithMode(ctx context.Context, inner net.Listener, mode ListenMode, options ...ListenOption) (_ net.Listener, err error) {
+	m := mode.get()
+	source := m.source
+	if source == nil {
+		source, err = workloadapi.NewX509Source(ctx, m.options...)
+		// Close source if there is a failure after this point
+		defer func() {
+			if err != nil {
```

Comment:
While writing unit tests I realized that we should check for `source != nil` also.

```suggestion
			if err != nil && source != nil {
```

If you agree, would you mind update it in dial as well?